Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,277 @@
+        EntityQuery<RadiationGridResistanceComponent> resistanceQuery)
+    {
+        var visitedTiles = new List<(Vector2i, float?)>();
+        var radRay = new RadiationRay(mapId, sourceUid,sourceWorld,
+            destUid,destWorld, incomingRads)
+        {
+            Grid = gridUid,
+            VisitedTiles = visitedTiles
+        };
+
+        // if grid doesn't have resistance map just apply distance penalty
+        if (!resistanceQuery.TryGetComponent(gridUid, out var resistance))
+            return radRay;
+        var resistanceMap = resistance.ResistancePerTile;
+
+        // get coordinate of source and destination in grid coordinates
+        // todo: entity queries doesn't support interface - use it when IMapGridComponent will be removed
+        if (!TryComp(gridUid, out IMapGridComponent? grid))
+            return radRay;
+        var sourceGrid = grid.Grid.TileIndicesFor(sourceWorld);
+        var destGrid = grid.Grid.TileIndicesFor(destWorld);
+
+        // iterate tiles in grid line from source to destination
+        var line = Line(sourceGrid.X, sourceGrid.Y, destGrid.X, destGrid.Y);
+        foreach (var point in line)
+        {
+            (Vector2i, float?) visitedTile = (point, null);
+            if (resistanceMap.TryGetValue(point, out var resData))
+            {
+                radRay.Rads -= resData;
+                visitedTile.Item2 = radRay.Rads;
+            }
+
+            // save data for debug
+            if (saveVisitedTiles)
+                radRay.VisitedTiles.Add(visitedTile);
+
+            // no intensity left after blocker
+            if (radRay.Rads <= MinIntensity)
+            {
+                radRay.Rads = 0;
+                return radRay;
+            }
+        }
+
+        return radRay;
+    }
+
+    private RadiationRay Raycast(MapId mapId, EntityUid sourceUid, EntityUid destUid,
+        Vector2 sourceWorld, Vector2 destWorld, Vector2 dir, float distance, float incomingRads,
+        EntityQuery<RadiationBlockerComponent> blockerQuery)
+    {
+        var blockers = new List<(Vector2, float)>();
+        var radRay = new RadiationRay(mapId, sourceUid, sourceWorld,
+            destUid, destWorld, incomingRads)
+        {
+            Blockers = blockers
+        };
+
+        var colRay = new CollisionRay(sourceWorld, dir, (int) CollisionGroup.Impassable);
+        var results = _physicsSystem.IntersectRay(mapId, colRay, distance, returnOnFirstHit: false);
+
+        foreach (var obstacle in results)
+        {
+            if (!blockerQuery.TryGetComponent(obstacle.HitEntity, out var blocker))
+                continue;
+
+            radRay.Rads -= blocker.RadResistance;
+            blockers.Add((obstacle.HitPos, radRay.Rads));
+
+            if (radRay.Rads <= MinIntensity)
+            {
+                radRay.Rads = 0;
+                return radRay;
+            }
+        }
+
+        return radRay;
+    }
```

Comment:
Won't this cause inconsistent behavior if using it with multiple grids?

1. For stuff like thindows with gridcasts you assume the whole tile is blocked, whereas with this thindows would get ignored.
2. The corner cutting mentioned elsewhere won't happen with raycasts.

You should always use the gridcast.

Reply:
1. Directional windows don't have radiation blockers. This should be added latter In future PR, similar how explosion system do it
2. Cross grid gridcast would be great, but it's not trivial to implement. In future iterations I would prefer to remove raycast completely

Reply:
As it is right now if someone snips a tile near singularity for example then suddenly radiation will function differently which is a bit sussy.

Reply:
Yes, that's true. But I honestly don't know how to do cross grid gridcast fast

Reply:
I'm fine with just having physics-raycast being used initially until a more consistent multi-grid cast algorithm gets added in some later PR, but on that note:

Wouldn't it actually be relatively straightforward to grid cast across multiple grids? The final radiation count only depends on the distance & sum of blockages encountered, so the order in which you process blockers/grids doesn't matter. So you could just perform the same grid cast on each grid individually? Though it could also be slightly sped up by skipping parts of the cast that are far outside of a grid's bounds by doing some Box2 - line intersection to get cropped start & end points for each grid?

Reply:
I implemented @ElectroJr suggestion and it worked really well. 

Combined gridcast + raycast approach was replaced by cross grid gridcast. I updated gridcast info and benchmarks in PR header.